File: Leetcode-Python/Leetcode-12 Integer to Roman.py
class Solution(object):
    def intToRoman(self, num):
        """
        :type num: int
        :rtype: str
        """
        res=""
        i=0
        dic={1:"I",5:"V",10:"X",50:"L",100:"C",500:"D",1000:"M"}
        while num>0:
            if num%10!=4 and num%10!=9:
                res=(num%10)*dic[10**i]+res if num%10<5 else dic[5*10**i]+(num%10-5)*dic[10**i]+res
            elif num%10==4:
                res=dic[10**i]+dic[5*10**i]+res
            else:
                res=dic[10**i]+dic[10*10**i]+res
            num/=10
            i+=1
        return res
#简单描述即可，注意一些细节：1.正着添加还是逆着添加 2.啥时候添加条件不同

File: Leetcode-C++/Leetcode-12 Integer to Roman.cpp
#include <vector>
#include <unordered_map>
#include <string>
using namespace std;

class Solution {
public:
    string repeat(const string& s, int n) {
            string result;
            for (int i = 0; i < n; i++) {
                result += s;
            }
            return result;
        }

    string intToRoman(int num) {
        unordered_map<int, string> dict = {
            {1, "I"}, {5, "V"}, {10, "X"}, {50, "L"}, {100, "C"}, {500, "D"}, {1000, "M"}
        };
        string res = "";
        int temp, base = 1;
        while (num > 0) {
            temp = num % 10;
            if (temp != 4 && temp != 9) {
                if (temp < 5) {
                    res = repeat(dict[base], temp) + res;
                } else {
                    res = dict[base * 5] + repeat(dict[base], temp - 5) + res;
                }
            } else if (temp == 4) {
                res = dict[base] + dict[base * 5] + res;
            } else {
                res = dict[base] + dict[base * 10] + res;
            }
            num /= 10;
            base *= 10;
        }
        return res;
    }
};